contracts/contracts/validator-registry/middleware/MevCommitMiddleware.sol

```diff
@@ -0,0 +1,704 @@
+        uint256 len = vaults.length;
+        for (uint256 i = 0; i < len; ++i) {
+            _deregisterVault(vaults[i]);
+        }
+    }
+
+    /// @notice Registers validators via their BLS public key and vault which will secure them.
+    /// @dev This function is callable by any delegated operator on behalf of a vault.
+    /// @param blsPubkeys BLS public keys of the validators to register.
+    /// @param vaults Addresses of vaults which will secure groups of validators.
+    function registerValidators(bytes[][] calldata blsPubkeys,
+        address[] calldata vaults) external whenNotPaused onlyValidBLSPubKeys(blsPubkeys) {
+        uint256 vaultLen = vaults.length;
+        require(vaultLen == blsPubkeys.length, InvalidArrayLengths(vaultLen, blsPubkeys.length));
+        address operator = msg.sender;
+        _checkOperator(operator);
+        for (uint256 i = 0; i < vaultLen; ++i) {
+            address vault = vaults[i];
+            uint256 keyLen = blsPubkeys[i].length;
+            _checkVault(vault);
+            uint256 potentialSlashableVals = _potentialSlashableVals(vault, operator);
+            require(keyLen <= potentialSlashableVals,
+                ValidatorsNotSlashable(vault, operator, keyLen, potentialSlashableVals));
+            for (uint256 j = 0; j < keyLen; ++j) {
+                _addValRecord(blsPubkeys[i][j], vault, operator);
+            }
+        }
+    }
+
+    /// @notice Requests deregistrations for validators, restricted to contract owner,
+    /// or the (still registered and non-blacklisted) operator of the validator pubkey.
+    /// @param blsPubkeys BLS public keys of the validators to request deregistrations for.
+    function requestValDeregistrations(bytes[] calldata blsPubkeys) external whenNotPaused {
+        uint256 len = blsPubkeys.length;
+        for (uint256 i = 0; i < len; ++i) {
+            _requestValDeregistration(blsPubkeys[i]);
+        }
+    }
+
+    /// @dev Deletes validator records, callable by contract owner,
+    /// or the (still registered and non-blacklisted) operator of the validator pubkey.
+    /// @notice This function allows the contract owner to combat a greifing scenario where an operator
+    /// registers a validator pubkey that it does not control, own, or otherwise manage.
+    function deregisterValidators(bytes[] calldata blsPubkeys) external {
+        uint256 len = blsPubkeys.length;
+        for (uint256 i = 0; i < len; ++i) {
+            _deregisterValidator(blsPubkeys[i]);
+        }
+    }
+
+    /// @dev Slashes validators and marks them for deregistration.
+    /// @param blsPubkeys The L1 validator BLS public keys to slash.
+    /// @param infractionTimestamps The block.timestamps for blocks during which each infraction occurred.
+    function slashValidators(bytes[] calldata blsPubkeys, uint256[] calldata infractionTimestamps) external onlySlashOracle {
+        uint256 len = blsPubkeys.length;
```

Comment:
we could cache bytes[] calldata pubkeyArray = blsPubkeys[i], to reduce redundant reads

Reply:
Not sure I understand the comment here, as `blsPubkeys` is already a one-dimensional calldata array

Reply:
Fixed a comment, thank you

Reply:
See https://github.com/primev/mev-commit/pull/383/commits/2cd1b3d82415c92bd7fbc0c055ffe05c163f89b4, I'll try to use local vars as much as possible in solidity, will also see if there's a linter rule for this